File: java/maximum-product-of-word-lengths.java
// 318. 最大单词长度乘积
// https://leetcode-cn.com/problems/maximum-product-of-word-lengths/
package java;

class MaxProductOfWordLength {
    public int maxProduct(String[] words) {
        int len = words.length;
        int[] mask = new int[len];
        for (int i = 0; i < len; i++) {
            String word = words[i];
            int wordLen = word.length();
            for (int j = 0; j < wordLen; j++) {
                mask[i] |= 1 << (word.charAt(j) - 'a');
            }
        }
        int ans = 0;
        for (int i = 0; i < len - 1; i++) {
            for (int j = i + 1; j < len; j++) {
                if ((mask[i] & mask[j]) == 0) {
                    ans = Math.max(ans, words[i].length() * words[j].length());
                }
            }
        }
        return ans;
    }
}

File: python/maximum-product-of-word-lengths.py
# 318. 最大单词长度乘积 - Medium
# https://leetcode-cn.com/problems/maximum-product-of-word-lengths/

from typing import List
from itertools import product
from functools import reduce


def maxProduct(words: List[str]) -> int:
    mask = [reduce(lambda a, b : (a | 1 << (ord(b) - ord('a'))), word, 0) for word in words]
    return max(len(x[1]) * len(y[1]) for x, y in product(zip(mask, words)) if x[0] & y[0] == 0, default=0)